ext/openssl/openssl.c

```diff
@@ -5309,48 +5340,704 @@ PHP_FUNCTION(openssl_pkcs7_decrypt)
 
 /* }}} */
 
-/* {{{ proto bool openssl_private_encrypt(string data, string &crypted, mixed key [, int padding])
-   Encrypts data with private key */
-PHP_FUNCTION(openssl_private_encrypt)
+/* {{{ CMS S/MIME functions taken from PKCS#7 functions */
+
+/* {{{ proto bool openssl_cms_verify(string filename, int flags [, string signerscerts [, array cainfo [, string extracerts [, string content [, string pk7]]]], sigfile, encoding])
+   Verifies that the data block is intact, the signer is who they say they are, and returns the CERTs of the signers */
+PHP_FUNCTION(openssl_cms_verify)
 {
-	zval *key, *crypted;
-	EVP_PKEY *pkey;
-	int cryptedlen;
-	zend_string *cryptedbuf = NULL;
-	int successful = 0;
-	zend_resource *keyresource = NULL;
-	char * data;
-	size_t data_len;
-	zend_long padding = RSA_PKCS1_PADDING;
+	X509_STORE * store = NULL;
+	zval * cainfo = NULL;
+	STACK_OF(X509) *signers= NULL;
+	STACK_OF(X509) *others = NULL;
+	CMS_ContentInfo * cms = NULL;
+	BIO * in = NULL, * datain = NULL, * dataout = NULL, * p7bout  = NULL, *sigbio = NULL;
+	zend_long flags = 0;
+	char * filename;
+	size_t filename_len;
+	char * extracerts = NULL;
+	size_t extracerts_len = 0;
+	char * signersfilename = NULL;
+	size_t signersfilename_len = 0;
+	char * datafilename = NULL;
+	size_t datafilename_len = 0;
+	char * p7bfilename = NULL;
+	size_t p7bfilename_len = 0;
+	char * sigfile = NULL;
+	size_t sigfile_len = 0;
+	zend_long encoding = ENCODING_CMS;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "szz|l", &data, &data_len, &crypted, &key, &padding) == FAILURE) {
+	RETVAL_LONG(-1);
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pl|p!ap!p!p!p!l", &filename, &filename_len,
+				&flags, &signersfilename, &signersfilename_len, &cainfo,
+				&extracerts, &extracerts_len, &datafilename, &datafilename_len,
+				&p7bfilename, &p7bfilename_len,
+				&sigfile, &sigfile_len,&encoding) == FAILURE) {
 		RETURN_THROWS();
 	}
+
+	in = BIO_new_file(filename, PHP_OPENSSL_BIO_MODE_R(flags));
+	if (in == NULL) {
+	  php_openssl_store_errors();
+	  goto clean_exit;
+	}
+	if ( sigfile && CMS_DETACHED ) {
+	  sigbio = BIO_new_file(sigfile, PHP_OPENSSL_BIO_MODE_R(flags));
+
+	  switch (encoding) {
+	  	case ENCODING_PEM:
+			cms = PEM_read_bio_CMS(sigbio,NULL,0,NULL);
+			datain = in;
+			break;
+	     	case ENCODING_DER:
+			cms = d2i_CMS_bio(sigbio,NULL);
+			datain = in;
+			break;
+	    	case ENCODING_CMS:
+	       		cms = SMIME_read_CMS(in, &datain);
```

Comment:
shouldn't this get `sigbio`?

Reply:
I don't think so.  In this case, cms is a struct *CMS_ContentInfo, in is the message, and datain is the content that is fed to CMS_verify.

Reply:
That doesn't seem like what's done for PEM and DER which takes sigbio as CMS_ContentInfo and in are just the data that is signed. Basically in this case you would completely ignore sigfile which I don't think should be the case.

Could you please also add a test for detached signatures?

Reply:
@bukka thanks again for your comments.   I was initially looking at the wrong code line the first time around.  

I think the code there is wrong.  That is- to my knowledge there is no way to output a detached signature in CMS form, and openssl natively does not provide a way to read just a CMS-encoded signature.  I therefore propose to produce an error. 

WILCO on the detached signature test case (this was actually impetus for doing this code!).

Reply:
I have done just this in the latest commits.

Reply:
@bukka Just to follow up on this, I had an email conversation with the co-author of CMS (Russ Housley) and the person who did the CMS code for Ruby, and they both agree with issuing an error.